File: C/Quick-Sort.c
#include <stdio.h>


int swap(int *pos1, int *pos2) {
    int temp = *pos1;
    *pos1 = *pos2;
    *pos2 = temp;
}


int print(int *arr, int len) {
    for(int i=0; i<len; ++i)
        printf("%d ", arr[i]);
    printf("\n");
}


int partition(int *arr, int low, int high) {
    int pivot = arr[high];
    int i = (low-1);

    for(int j=low; j<high; ++j) {
        if(arr[j] <= pivot) {
            ++i;
            swap(&arr[i], &arr[j]);
        }
    }
    
    swap(&arr[++i], &arr[high]);

    return i;
}


void sort(int *arr, int low, int high) {
    if(low < high) {
        int p = partition(arr, low, high);
        sort(arr, low, p-1);
        sort(arr, p+1, high);
    }
}


int main() {
    int arr[6] = {10, 7, 8, 9, 1, 5};
    int len = 6;

    sort(arr, 0, len-1);
    print(arr, len);
}

File: Python/Quick-Sort.py
from random import randint


def partition(arr, low, high):

    # pivot (Element to be placed at right position)
    pivot = arr[high]

    i = (low - 1)  # Index of small element

    for j in range(low, high):
        # If current element is smaller than
        # or equal to pivot
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i+1], arr[high] = arr[high], arr[i+1]
    print(arr)
    return i+1


def sort(arr, low, high):
    if low < high:
        # pi is partitioning index, arr[p] is now
        # at right place
        pi = partition(arr, low, high)
        sort(arr, low, pi-1)
        sort(arr, pi+1, high)


def main():
    arr = [5, 8, 1, 3, 7, 9, 2]  # [randint(0, 100) for _ in range(10)]
    length = 7  # 10
    print(arr)
    sort(arr, 0, length-1)
    print(arr)


if __name__ == '__main__':
    main()
